### ISSUE TITLE
Graceful shutdown

### ISSUE THREAD

**Issue by matreyes** [kafka4beam/brod#442]
Hi guys. 
I think I'm doing something wrong, seems that I don't get the supervision strategy.
I have a pipeline that parses and sends all messages to a new topic. 
When I shut down the server, some messages are sent but not commited, which causes duplication (I need exactly once delivery).
Here is an example code (I'm sorry, it's in elixir, but could be helpful for other folks).

```elixir
defmodule TestPipeline do
  @behaviour :brod_group_subscriber_v2
  @source :test_consumer

  require Logger

  def child_spec(_opts) do
    %{
      id: __MODULE__,
      start: {__MODULE__, :start, []},
      type: :worker,
      restart: :permanent,
      shutdown: 10_000
    }
  end

  def start() do
    group_config = [
      offset_commit_policy: :commit_to_kafka_v2,
      offset_commit_interval_seconds: 2,
      rejoin_delay_seconds: 3,
      reconnect_cool_down_seconds: 10
    ]

    config = %{
      client: @source,
      group_id: "test",
      topics: ["test"],
      cb_module: __MODULE__,
      group_config: group_config,
      consumer_config: [begin_offset: :earliest]
    }

    :brod.start_link_group_subscriber_v2(config)
  end

  def init(_arg, _arg2) do
    {:ok, []}
  end

  def handle_message(messages, _state) do
    IO.inspect("processing message #{inspect(messages)}") # here I send the same batch to the new topic.
    :timer.sleep(3_000)
    IO.inspect("commiting offset")
    {:ok, :commit, nil}
  end
end
```

and my test:
```elixir
  test "shutdown gracefully" do
    brokers = [{:kafka, 9092}]

    :ok = :brod.start_client(brokers, :test_consumer)

    {:ok, pid} = start_supervised(TestPipeline)
    :timer.sleep(1_000)
    :init.stop()
    IO.inspect("shutting down")
    :timer.sleep(20_000)
  end
```

And it shows:
```
"processing message .... (always same messages)"
"shutting down"
```
But the offset is not commited

Thanks!

**Comment by zmstone** [comment]
Hi @matreyes 
Exact once delivery is not supported in brod.
offset commits are collected in batches with delay and compacted before sending to Kafka
this is due to the fact that Kafka will not be able to handle per message or even per batch commits

It does not matter how you fiddle with the code here, there is always a risk of duplication.
e.g. even if the commit is synced, it may get lost due to a network disturbance
e.g. producer may fail to produce a batch and then retry, which will cause duplication in Kafka i.e. even Kafka offset is not trustworthy as a unique identifier for your messages.

There are a few primitive transactional APIs implemented in [kafka_protocol](https://github.com/kafka4beam/kafka_protocol)
which might be able to help achieving exact once delivery.
There are a lot of work to be done to make use of these transaction APIs, 
It requires both the producer and consumer to use transaction APIs for exact once delivery.

And IMHO, it's actually easier to implement at application level: 
Producer add sequence (or at least monotonically incasing) numbers for each message (per partition). 
And consumer commits its state on disk (or db), then drop duplicates if received again

**Comment by zmstone** [comment]
Hi @matreyes fair points.
I have sent https://github.com/kafka4beam/brod/pull/447 to hopefully address this issue. 
Will add some tests, but before that, it would be great if you can test it a bit in your setup.

About the transactional APIs in kafka_protocol,
the main risk, I believe, is that we have not used it in any production env.
Also, it's been a long time, I have pretty much lost all the context in my head,
it'll take some time for me to get up to speed again in case of issue reports and/or questions etc.

**Comment by matreyes** [comment]
Hi @zmstone 

It did work as expected! Great work!
(just look at https://github.com/zmstone/brod/commit/1fad7df863573ec8fa113b75716aa645bb6de743#diff-71a29fd4f3d8821238e03f7818172ca558c16233e79692b2b7afdcb56f737b92R398 ), it has more arguments than interpolated)

Best

**Comment by zmstone** [comment]
Thanks.
I'll fix the logging and add more tests.

**Comment by zmstone** [comment]
fix released in 3.15.3, published to hex.
